### PR TITLE
Support building dockerfile & fix Terraform PR args

### PR DESCRIPTION
**Pull request by bondezbond**
Solves #230 #232 

## AgentNodePlugin
* Added ability to build using `Dockerfile`
* Added ability to pass build-args

## TerraformPlanResultsPR
* Use `withArgument` to pass terraform arguments

### CODE REVIEW

**Comment by bondezbond** [comment]
@jleopold28 yup tested on local repo

**Comment by dkolb** [comment]
Can we perhaps make this another method and accept an optional path to the docker file?

```
AgentNodePlugin.withAgentDockerImage('local/custom-docker-image:latest')
    .withDockerfile(true)
    .withDockerfilePath('path/to/Dockerfile'')
```
This would be a bit more flexible to various repo layouts and stick with our more descriptive option naming instead of just a `true` sitting at the end of a method call.

File: docs/AgentNodePlugin.md [hunk 1]
```diff
@@ -2,13 +2,40 @@
...
 
+### Using pre-build docker image
 ```
 // Jenkinsfile
 @Library(['terraform-pipeline']) _
 
 Jenkinsfile.init(this)
 
-AgentNodePlugin.withAgentDockerImage('hashicorp/terraform:0.10.2')
+def terraformVersion = TerraformPlugin.checkVersion()
```
Comment (kmanning) [inline docs/AgentNodePlugin.md:12]: Is this pertinent to the AgentNodePlugin?

File: docs/AgentNodePlugin.md [hunk 2]
```diff
@@ -2,13 +2,40 @@
...
+```
+
+### Using Dockerfile in repo
+```
+// Jenkinsfile
+@Library(['terraform-pipeline']) _
+
+Jenkinsfile.init(this)
+
+def terraformVersion = TerraformPlugin.checkVersion()
```
Comment (kmanning) [inline docs/AgentNodePlugin.md:36]: Same as above - is terraformVersion actually pertinent to the AgentNodePlugin?

File: src/TerraformPlanResultsPR.groovy [hunk 3]
```diff
@@ -44,10 +44,12 @@ class TerraformPlanResultsPR implements TerraformPlanCommandPlugin, TerraformEnv
     @Override
     public void apply(TerraformPlanCommand command) {
         if (landscape) {
```
Comment (kmanning) [inline src/TerraformPlanResultsPR.groovy:46]: Building landscape logic into this plugin is continuing to feel "wrong".  This PR is already uncovering that there's a weakness in TerraformCommands, and it's starting to address the weakness (by pushing `-out=tfplan` into an argument, where it belongs).  I think we need to take this further, and maybe start spelling out a `.withOutputRedirection` or something similar, for `2>plan.err` to live.  

If the stderr redirection was treated differently than a normal "suffix" (and as you're illustrating, it is different), then we could push the landscape logic out of this plugin as well.

^-- this doesn't need to be done as part of this particular PR, but it should be done before the PR Plugin is released.

File: src/AgentNodePlugin.groovy [hunk 4]
```diff
@@ -36,10 +44,14 @@ public class AgentNodePlugin implements TerraformValidateStagePlugin, TerraformE
 
     public Closure addAgent() {
         return { closure ->
-            if (dockerImage) {
+            if (this.dockerImage && this.withDockerfile == false) {
                 docker.image(this.dockerImage).inside(this.dockerOptions) {
                     closure()
                 }
+            } else if (this.dockerImage && this.withDockerfile == true) {
```
Comment (kmanning) [inline src/AgentNodePlugin.groovy:51]: It's minor, but `withDockerfile == true`, `withDockerfile == false` feels wonky.  

You could collapse this logic into:
```
if (dockerImage) {
    if (withDockerfile) {
        docker.build(...)
    } else {
        docker.image(...)
    }
} else {
    ....
}
```
Or at the very least, not compare against the literals:
```
if (this.dockerImage && !this.withDockerfile) {
    ...
} else if (this.dockerImage && this.withDockerfile) {
    ...
}
```

**Comment by kmanning** [comment]
Can you update the ChangeLogs to reference your change?

File: src/AgentNodePlugin.groovy (hunk 4 above)
Comment (kmanning) [reply src/AgentNodePlugin.groovy:51]: Ah, I think this is why it feels wonky to me - this is  in the Sun Coding Standard - https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/sun_checks.xml#L171

**Comment by kmanning** [comment]
Related to #193

File: docs/AgentNodePlugin.md (hunk 1 above)
Comment (bondezbond) [reply docs/AgentNodePlugin.md:12]: that isn't pertinent. I can remove that feature but I wanted to show. how to use use it with the terraform version. Cause just having a terraform version file won't pull the right image.

File: docs/AgentNodePlugin.md (hunk 2 above)
Comment (bondezbond) [reply docs/AgentNodePlugin.md:36]: same response as above

File: src/AgentNodePlugin.groovy (hunk 4 above)
Comment (bondezbond) [reply src/AgentNodePlugin.groovy:51]: I can change it either way. I tend to write literal checks so you know exactly what the code is looking for

**Comment by bondezbond** [comment]
@dkolb great suggestion. I'll update that option. Maybe I'll just do
```
    .withDockerfilePath('path/to/Dockerfile'')
```
instead of two functions? Cause to me if your passing DockerFilePath inherently that would mean `withDockerFile = true`, does that sound good?

**Comment by bondezbond** [comment]
thanks for the review @kmanning i'll update everything you mentioned

**Comment by dkolb** [comment]
@bondezbond Yeah that sounds perfect.

**Comment by kmanning** [comment]
This PR was replaced by https://github.com/manheim/terraform-pipeline/pull/241.  Closing this without merging.

**Pull request closed without merging**